project_generator/workspace.py

```diff
@@ -37,107 +38,195 @@ def __init__(self, projects_file='projects.yaml', project_root='.'):
             self.settings.update(self.projects_dict['settings'])
 
         # so that we can test things independently of eachother
-        self.projects = {}
-
+        self.workspaces = {}
+
+        # We support grouping of projects or just a project for ProjectWorkspace
+        #
+        # [projects.yaml]
+        #   projects:
+        #       project_group_1:
+        #           project_1:
+        #               -a
+        #               -b
+        #               -c
+        #           project_2:
+        #               -d
+        #               -e
+        #               -f
+        #       project_3:
+        #           -g
+        #           -h
         if 'projects' in self.projects_dict:
-            for name,records in self.projects_dict['projects'].items():
-                if "common" in records:
-                    self.projects[name] = Project(name, records, self)
+            for name, records in self.projects_dict['projects'].items():
+                if type(records) is dict:
+                    # workspace
+                    projects = [Project(n, uniqify(flatten(r)), self) for n, r in records.items()]
                 else:
-                    x = set([item if len(item)>1 else sublist for sublist in records for item in sublist])
-                    self.projects[name] = Project(name, list(x), self)
+                    # single project
+                    projects = [Project(name, uniqify(flatten(records)), self)]
+
+                self.workspaces[name] = ProjectWorkspace(name, projects, self, type(records) is not dict)
         else:
             logging.debug("No projects found in the main record file.")
+        # extension - workspaces
+        if 'workspaces' in self.projects_dict:
```

Comment:
Should workspace take precedent over projects? Two if's means two passes through project list.
Perhaps:
```python
if 'workspaces' in self.projects_dict:
      blah
elif 'projects' in self.projects_dict:
      blah
```